### PR TITLE
fix Issue 14940: Can't call logger with more complex objects

### PR DESCRIPTION
**Pull request by burner**
https://issues.dlang.org/show_bug.cgi?id=14940

### CODE REVIEW

**Comment by Hackerpilot** [comment]
:white_check_mark:

File: std/typecons.d [hunk 1]
```diff
@@ -1799,8 +1799,7 @@ Params:
         _isNull = false;
     }
 
-    template toString()
-    {
+    template toString() {
```
Comment (quickfur) [inline std/typecons.d]: Nitpick: opening brace on own line, please.

**Comment by burner** [comment]
@quickfur @Hackerpilot done

**Comment by mathias-baumann-sociomantic** [comment]
While this fixes the specific issue, isn't the deeper problem that `@safe` is required at all?

**Comment by burner** [comment]
@mathias-baumann-sociomantic not really the OutputRange delegate passed to Nullable is not @safe and as this delegate is called by toString it can not be @safe

**Comment by mihails-strasuns-sociomantic** [comment]
Code duplication looks a bit worrying here. Can't dmd infer attributes for delegate here? (considering `toString` is already a template)

**Comment by burner** [comment]
@mihails-strasuns-sociomantic no it didn't making the delegate a template parameter also didn't work. The alias this was then used instead of the toString

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by DmitryOlshansky** [comment]
Puttin aside the general problem with delegates, merging this

**Comment by burner** [comment]
thanks

**Pull request merged**